### ISSUE TITLE
Workaround for logout in the Xplora® App

### ISSUE THREAD

**Issue by Ludy87** [Ludy87/xplora_watch#24]
1. add phone number from watch to contacts (as admin)
2. create a login profile for this phone number
3. log into the app with the created profile to test whether it works
4. use the login data for the integration

**Comment by FerretMonger** [comment]
When would I need to do this? I just got the watch and want to set the integration for the first time.

**Comment by Ludy87** [comment]
Use the Xplora app to find the contact's number as described above.

**Comment by mschiessl** [comment]
They only allow "2" administrators, which is me and my wife - so I cannot add the watch itself as an admin.
Any suggestions? 
Should we raise a support on xplora side ?

**Comment by Ludy87** [comment]
@mschiessl create an account with your watch's phone number and include the number as a contact. The integration does not require an admin as a user.

**Comment by mschiessl** [comment]
Thanks, @Ludy87, makes total sense now! 
I  got the following sentence totally wrong:
 
```
add phone number from watch to contacts (as admin)
```

could be better:

```
As an admin user, add phone number from watch to contacts
```

**Comment by P-Stenbumling** [comment]
> @mschiessl create an account with your watch's phone number and include the number as a contact. The integration does not require an admin as a user.

How and where do you create an account? 
from the mobile app i'm unable to get an verification, have added the xploras number to contacts.

**Comment by Ludy87** [comment]
1. Download the app [Android](https://play.google.com/store/apps/details?id=com.xplora.xplorao2o) or [Apple](https://apps.apple.com/ de/app/xplora/id1521983878).
2. You select the item "Create account"
3. Follow the instructions
4. Done

---
You are still logged into the app:
1. log out
2. You select the item "Create account"
3. Follow the instructions
4. Done

**Comment by xeor** [comment]
I'm a little confused by this workaround..

* Is it needed only if you have problems that you are logged out, or is it always needed?
* Is the goal to have a dedicated login user for the API?
* What if you have multiple watches, will 1 watch be used for both numbers and need access to both?
* Will the `phonenumber` setting be the same as `child_phonenumber`?
* Will the kid have a "dummy" number (itself) on the watch?

Don't need answer to all of it, but something to shed some light on the confusion.. It seams to be multiple issues that have problems with login.

**Comment by Ludy87** [comment]
An admin/user/parent can only login on ONE device, either in the app on the mobile phone or in the integration. It is therefore necessary to create an app account on the mobile phone in order to be able to use an "extra" user for the integration.

`phonenumber` is the number from App/integration User, not from watch.

The watch usually has its own phone number, hence its own key 'child_phonenumber'
In itself, the watch has a unique ID ('watch_id')
https://github.com/Ludy87/xplora_watch/blob/b97df72bc152630dafc6f8efc4792aabe2d6df3a/custom_components/xplora_watch/__init__.py#L135

**Comment by xeor** [comment]
Does this trick work anymore? When logged in as a parent, I can see the watch have it's own phone-number as contact. I can also login using the same phone-number in the app and verify that it sees itself.

* 111: Phone-number for the simcard on the watch, using the same as was installed when bought
* 222: My own (parent) phone-number

I removed, password, timezone, userlang and country_code below in the configuration examples

```yaml
# This gives `pyxplora_api.exception_classes.LoginError: Login to Xplora® API failed. Check your input! - 1`
xplora_watch:
  phonenumber: "111"
  child_phonenumber:
    - "111"

# This works, but throws me out from the app on the phone when home-assistant is starting
xplora_watch:
  phonenumber: "222"
  child_phonenumber:
    - "111"
```

I also tried "manually" using the code found in `pyxplora_api`.

```py
# This works
gql.runGqlQuery(gql_mutations.SIGN_M.get("issueTokenM", ""), {'countryPhoneNumber': '+47', 'phoneNumber': '222', 'password': 'password-hash-here', 'userLang': 'nb-NO', 'timeZone': 'Europe/Oslo'})

# same with phoneNumber of watch gives
{'errors': [{'code': 'E000025', 'message': 'Incorrect password.'}], 'data': {'issueToken': None}}
```

I got `E000025` on my own phone-number in the beginning as well, however, that all the sudden fixed itself after fiddling with resetting the password on the account (even tho I could login using the password I tried)..

Any tips?

**Comment by Ludy87** [comment]
@xeor  Has a user account been created with the phone number from the watch?

**Comment by xeor** [comment]
@Ludy87 yes, I have created one from the xplora app and I can login using it. I see in it's profile-page that it has a dedicated mail-address as well. It can see the watch when logged in, but there is no information like location and such (since it's not a guardian?).

I have also tried to remove the 111 number from the contact-list and readd it using the qr code instead of adding it using guardian>add-contact. Both methods yields the same error..

**Comment by Ludy87** [comment]
Try the minimal configuration!

https://github.com/Ludy87/xplora_watch#minimal-configuration

```yaml
xplora_watch:
  country_code: "+49"
  phonenumber: "151XXXXXXXX"
  password: "password"
  userlang: "de-DE"
  timezone: "Europe/Berlin"
```

**Comment by xeor** [comment]
@Ludy87 same error as before..

I also tried to change the password on the 111 number and confirming that it works to login using the app using 111.
Retried everything with the new password, but still a no-go, same error :(

**Comment by Ludy87** [comment]
@xeor I ran some tests and came to the conclusion: You are using the wrong password. Please check if all characters are correct. No hash value, no additional characters such as spaces.

**Comment by xeor** [comment]
Thanks for checking! I will do some more checks tomorrow. I'm copy-pasting the password from where it works into the config, and I have made sure it doesnt contain any strange characters. I even had the same password on both accounts to test and verified every login (outside home-assistant) worked..

I'll come back with some more tests :)

**Comment by xeor** [comment]
@Ludy87 ! I finally got it to work! Very useful the new version I installed from HACS with better errors. :)
It was the account-creation that wasn't correct. I thought I had created an account on the phone-number, but what I had created was an account on a separate mail, and attached the phone-number..

In the app, way I ended up doing it was (I think); "Create an Account" > "Skip" > "Log in" > "Forgot password" > "Phone number" > the 111 number > ...

Before this, I was able to login using the account on myxplora and in the app, but the account was still not correctly created using the phone-number.. So for the next; use the phone-number! Don't invent a new mail-address

**Comment by Aunion** [comment]
I was having similar issues with this process as @xeor and I can confirm that using the "forgot password" process (using my phone number as login) also did the trick for me (curiously the service sends the phone number verification code to your email, waited for quite a bit for it to arrive as sms before noticing that...)

**Comment by jonathanbower** [comment]
How can I create an account without an email adress? When selecting "create account" i only get an option to enter email and password. I if i select forgot account and enter child_phonenumber I get "User doesn't exist".

**Comment by Aunion** [comment]
@jonathanbower I'm pretty sure you have to create one with an email, then attach a phone number to it, and then reset the password using that phone number. At least those were the steps I had to take to get this to work.

**Comment by jonathanbower** [comment]
@Aunion , that did the trick, thanks!

**Comment by westis** [comment]
I don't really understand this.

1. When logged in with my admin account in the Xplora app, add my child as a contact to himself.
2. Log out and create an account for my child, with his email address, but my phone number as the main contact.
3. Then I use what login data in the HA integration? The description above just says "Use the login data". But in a later comment it's like the phonenumber and child_phonenumber in configuration.yaml should not be the same? 

Not sure if it has to do with the above or something else, but "last tracking" hasn't updated for me for more than an hour now, and battery level is -1.

**Comment by xeor** [comment]
The  -1 might be another problem.. I just noticed the smell of thing on my setup, but didn't investigate yet

**Comment by eirikgrindevoll** [comment]
This doesnt seem to work. What actually worked was that one of my sons friends watch became avalible! And i got all the information from that watch! Because they have added eachother as contacts

**Comment by eirikgrindevoll** [comment]
> This doesnt seem to work. What actually worked was that one of my sons friends watch became avalible! And i got all the information from that watch! Because they have added eachother as contacts

Ok so managed to get it to work. For me i had to remove the contact on my admin account and then re add the contact (that is used for authentication to the API) Since my sons number is added on his friends account we can still view information from that watch not good, but this workaround wouldnt work if not.

BUT now i need help. When i removed the addon from HACS i also cleaned up the entities ect. And if i recall also cleaned up known_devices.yaml? I do not manage to get a sensor now for tracking the device?!

**Comment by Ludy87** [comment]
@eirikgrindevoll can you also read your friend's watch in the app? That would be a massive data breach.

**Comment by eirikgrindevoll** [comment]
> @eirikgrindevoll can you also read your friend's watch in the app? That would be a massive data breach.

It shows in the app, but only chat and call feature. But API shows everything that i also gets from my sons watch (right now no tracking data because i think i have messed something up in HA) . In app i also can choose "remove watch" and then it also removed my son from friends contacts. I havent tested with creating an account for another approved contact. Could test with grandparents to see if its actuly because account is created on my sons phonenumber. Would still call it a data breach yes. When you approve a contact you belive its only so that person can call the watch, not monitor it. But this whole workaround is based on that.

**Comment by sevorl** [comment]
I agree, this is a data breach in the Xplora API. But, as soon as Xplora would fix this, this integration would no longer work for most  scenarios. 
I would assume only Admins should be allowed to track location. But xplora limits the number of admins to two (typlically mum and dad).
Home Assistant would have to be an additional Admin.

**Comment by eirikgrindevoll** [comment]
They could allow for admins to sign in to multiple sessions. Lets say if you want to have the app on both your phone and tablet.

For my issue with integration not tracking the device i would love if someone had any idea. I cant remeber but i think i removed everything and also deleted the entry from known_devices.yaml? I have been searching the HA DB if there is something left since i have setup the itegration 2 times now and that makes it not create the device tracker entity?

**Comment by eirikgrindevoll** [comment]
I finaly got the device tracking working. I updated from august to september HA update and then it started working. I belive there might be an issue were if you remove the custom integration something is left over in HA but i cannot figure out what. Even deleted the DB didnt do the trick, but updatating HA core worked. Strange

**Comment by th00mas88** [comment]
> I finaly got the device tracking working. I updated from august to september HA update and then it started working. I belive there might be an issue were if you remove the custom integration something is left over in HA but i cannot figure out what. Even deleted the DB didnt do the trick, but updatating HA core worked. Strange

I have gone from last 3 ? HA and still the same issues. Did you remove Xplora from HACS  . update and start again ?   I am now on latest HA and core.

**Comment by CJA85** [comment]
Any idea if this workaround is viable for me? Both my kids have Xplora watches, can i map both of them like this in HA?

**Comment by Ludy87** [comment]
@CJA85 If they are both connected to the registered account, this should work.

**Comment by CJA85** [comment]
> @CJA85 If they are both connected to the registered account, this should work.

I have created another account with my work cell and added that number as a contact on my kids phones, but i cant seem to log in to the account through the integration, getting connection failed.

Can connect fine through my regular account, i tried reseting the password on the second account but that doesnt work.

Do i HAVE to create an account on one of the watches phonenumbers?

**Comment by sevorl** [comment]
No, you don't have to create an account for the watches phonenumbers. I use my work cell as well, and that works fine.

**Comment by eirikgrindevoll** [comment]
So this hit the news in Norway last week https://www.digi.no/artikler/kan-spore-andres-barneklokker-i-sanntid-uten-at-de-vet-om-det/522449 (Behind paywall) The Norwegian Data Protection Authority did NOT find it too break any personal data security. 

Seems like today they removed this workaround solution anyways? 
And still limits administrators to one sessions  and 2 administrators :( Too bad!

**Comment by Ludy87** [comment]
If they didn't continue to develop so much, then they could also close another sizeable security gap.

**Comment by Ludy87** [comment]
I like the link to my repository 👍

**Comment by Ludy87** [comment]
please check: https://github.com/Ludy87/xplora_watch/discussions/96

**Comment by smarthomefamilyverrips** [comment]
@Ludy87 does this still work? Or only works to login with admin account now?

**Comment by xsirpreisx** [comment]
I don't get this to work. 

- I added the phone number of my first sone to his own watches contacts (from the XPlora app on my phone, signed in with my own admin account).
- I then created an email account for him
- I logged out of my xplora app, and tried creating a new account using my sons email address. In the second step, I have to type in a phone number (probably his?). When I use his number, it says that "this number is already used" and I can't finish account creation.

I then tried it a second time, this time for my second son. 
- I also created a new email address.
- This time I didn't add his own number to his watches contacts
- Same thing when creating the account on the app, "this number is already" used.


I can go to "lost password", type in one of their phone numbers,  In the next step I can type in a password, twice. Submitting results in "Invalid request".


Can anyone help?

**Comment by smarthomefamilyverrips** [comment]
@xsirpreisx this method described in this post not works anymore, just use your own phone number and password you also use as guardian to login to the integration

**Comment by xsirpreisx** [comment]
But then I will get logged out on my phone?! Which I don't want. Am 12.03.2023 18:39 schrieb sesame_madness ***@***.***>:
@xsirpreisx this method described in this post not works anymore, just use your own phone number and password you also use as guardian to login to the integration

**Comment by smarthomefamilyverrips** [comment]
@xsirpreisx no you won't, I am logged in on my phone and on HA and also my wife is logged in as guardian with her own phone number

**Comment by BigFriendlyDuck** [comment]
Weird.

I’m being logged out.
And when I asked Xplora, that is how it should be.

I’ve tried asking Xplora about support for three admins or being logged in with three devices simultanously.
There can be many reasons for that, for an example if a kid is a part of two house-holds, and more than two adults is by all means a guardian.

So. Maybe more people should ask them? It’s needed for several reasons regardless of HA..

**Comment by smarthomefamilyverrips** [comment]
I not get logged out from anywhere, but I never have the xplora app open, I just log in and close it, but I receive notifications from app and if I open the app I am already logged in and not get logged out on HA nor app.... so not understand why you do.... anyway the above method not works anymore to my knowledge....

**Comment by xsirpreisx** [comment]
> @xsirpreisx no you won't, I am logged in on my phone and on HA and also my wife is logged in as guardian with her own phone number

I got logged out also when logging in via HA.

**Comment by smarthomefamilyverrips** [comment]
@xsirpreisx strange I not have this, and when you log back in again on app? It logs you out from HA?

**Comment by MrGlad8** [comment]
I have a problem with the no sensors being created after the second step with adding the integration. The login works and the Xplora-integration is created, but no sensors is there! Why is it doing this?

**Comment by Ludy87** [comment]
@MrGlad8 you have to select the appropriate sensors in the configuration.

**Comment by MrGlad8** [comment]
> @MrGlad8 you have to select the appropriate sensors in the configuration.

Ah ofc! And btw, should I login with my account that I use on the app, or with my Childs number that the watch is connected to?

**Comment by JoergBr** [comment]
> > @xsirpreisx no you won't, I am logged in on my phone and on HA and also my wife is logged in as guardian with her own phone number
> 
> I got logged out also when logging in via HA.

Same here. Followed the same process and got the same problems with "number is already used", also having the frequent logging out of the app with the adminstrator account (HA 2023.2.5).

**Comment by DIYtechie** [comment]
I cannot create a user with watches own number?

While being logged in as admin user on the App, I added the number of the sim card in the watch itself to the contacts on the watch. Then logged out and attempted to create a new user using the number. First of all, I could only create a new user by also using a new email, and when I typed in the watches own number I got the "number is already used".

**Comment by timn** [comment]
I have a similar experience of the two previous comments: always logged out of app (which is bad, because one cannot receive messages anymore) and registering with the watch's number is not possible. This effectively breaks the extension.

**Comment by otih** [comment]
I was not able to implement the workaround and keeping two admins active. Any plan/idea by when this issue may get solved?
As for now, this integration is broken.

**Comment by thmz82** [comment]
Unfortunately the workaround also doesn't work for me:
1. Start Xplora App on smartphone
2. "Create Account"
3. "Skip"
4. Enter a valid email address
5. Enter smart watches phone number -> "This mobile number is already in use. Please enter another mobile number."

Also tried to reset the password function
1. Start Xplora App on smartphone
2. Login
3. "Forgot password?"
4. Enter smart watches phone number 
5. Enter 2x new password and check "I am not a robot" -> "Invalid request"

Looks like they prohibit creating an account with the smart watches phone number.
Any other ideas on how to workaround that problem?

**Comment by Ludy87** [comment]
@thmz82 What do you want to do afterwards? If you want to fully use the integration, you have to log in with your admin account.

**Comment by thmz82** [comment]
> @thmz82 What do you want to do afterwards? If you want to fully use the integration, you have to log in with your admin account.

Primarily I would like to use the integration for tracking the GPS location and battery state. Also I'd like to enable/disable the school mode or alarms via automations (for example on holidays).
On the long run it would also be nice to fetch/send messages - but for now I'm using the Xplora App and it is sufficient.

The problem I currently have is that when I use my admin account in the HA integration, I get logged out from my mobile app. And when I log in again on my mobile phone, the HA integration gets in trouble (all sensors stop updating or get "unknown").
Are you familiar with this problem?

It's a pity that they allow us only to add 2 admin contacts. Already wrote to the support if there's a way (or maybe in the future) to add a 3rd one but no answer until now..

**Comment by Ludy87** [comment]
> The problem I currently have is that when I use my admin account in the HA integration, I get logged out from my mobile app. And when I log in again on my mobile phone, the HA integration gets in trouble (all sensors stop updating or get "unknown").
Are you familiar with this problem?

Yes, I know, there is no way to get around this behavior.

> It's a pity that they allow us only to add 2 admin contacts. Already wrote to the support if there's a way (or maybe in the future) to add a 3rd one but no answer until now..

I can tell you the answer: "Hi Ludy, thank you very much for the feedback & message, unfortunately it is currently only possible to appoint a total of 2 administrators."

**Comment by DIYtechie** [comment]
This issues seems to have been fixed by xplora. At least I can now be signed in via the app and the integration with the same (admin) user at the same time.

**Comment by Ludy87** [comment]
When I have a watch again I will test it. Thanks for the info.

**Comment by JanSchnacki** [comment]
Just another "works" comment.

I added the watch to HA using my own account (e-mail, not phone number). It works (after I understood how to add the entities) and I'm **not** logged out from the app (at least not for the last 30 minutes).

Thank you, Ludy87. This integration will help me a lot.

**Comment by otih** [comment]
I tested again today. After login with email it works great and I did not get logged out for several hours.

Seems this issue has been solved.

**Comment by Dan1jel** [comment]
This dose not work for me, tried both number and email, app on phone still says after a few min that's it I've been logged out due to someone logged in on a different device..

**Comment by mmarm** [comment]
Unfortunately for me it is also not possible to create an account for the watch Number since it is "already in use" and when trying to reset the password it says it can't find the user.. Also it does require to sign um with a number. Only an email doesn't work (anymore?!?).

Tried to also contact support and pitch in for more Guardians, since two is, even without the integration, too little. My grandparents (at least one of them) would need access. This limitation is drastically decreasing the value of this watch :-/

It is almost impossible to get a request through to the actual support. Hint for others wanting to increase demand for more Guardians: When you exhaust the chatbots answers you get the option to finally enter a support request. 

Still, the one that went through still came back with a generic "try this first" answer because the main topic I chose was "trouble with the app", so you have to resubmit AGAIN.
